File: papr/layouts/oneyear.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import math
import datetime
import logging

import cairo
from gi.repository import Pango
from gi.repository import PangoCairo

from util import metrics
from util import drawing


def drawCalendar(env):
    logging.debug(
        "Adding aditional information to enviroment specific to this cal style")
    env.font_size = 6
    env.line_width = 0.01 * metrics.CM
    env.row_width = (env.height - (13.0 * env.safety)) / 12.0
    env.row_height = (env.width - (2 * env.safety)) / 32

    logging.debug('row is %smm/%spx wide', env.row_width / metrics.MM, env.row_width)
    logging.debug('row is %smm/%spx high', env.row_height / metrics.MM, env.row_height)

    logging.debug("Creating Cario Surface and Context")
    logging.debug("width = %sp/%scm, height = %sp/%scm", env.height,
                  env.height / metrics.CM, env.width, env.width / metrics.CM)
    surface = cairo.PDFSurface(env.out, env.height, env.width)
    cr = cairo.Context(surface)

    date = datetime.date(env.year, env.month, 1)
    drawMonth(cr, env, date) # TODO using it totally worng, loop here over the function not in the function itself!
    logging.info("Finished drawing Calendar!")


def drawMonth(cr, env, date):
    # Creating a new date object with the first day of the month to draw
    logging.info("drawing %s...", date.strftime("%B %Y"))

    # Defining a one day timedelta object to increase the date object
    ONE_DAY = datetime.timedelta(days=1)
    for columnNo in range(0, 12):  # Iterate over 4 Month. 4 Month fit on one page
        with drawing.restoring_transform(cr):
            # move on the page to draw next month
            cr.translate(env.safety + (columnNo * env.row_width) +
                         (columnNo * env.safety), 0)
            drawMonthTitle(cr, env, date)
            # for every day of the month
            startingMonth = date.month
            while date.month == startingMonth:
                drawDay(cr, env, date)
                # increment date by one day
                date += ONE_DAY


def drawMonthTitle(cr, env, date):
    with drawing.restoring_transform(cr):
        with drawing.layout(cr) as layout:
            size = math.ceil(env.row_height * 0.9) # calculating font-size depending on the row height
            font = Pango.FontDescription("%s %s" % (env.fontHeading, size))
            layout.set_font_description(font)

            # preparing month string
            style = "%b"
            # TODO String is always abbreviated, adjust the font-size depending on the row width for all month titles!
            # style = "%B"
            # if(env.abbreviate_all):
            #     style = "%b"
            monthString = date.strftime(style)

            layout.set_text(monthString, -1)
            xOffset = (env.row_width - layout.get_pixel_size()[0]) / 2
            yOffset = (env.safety + (1.0 * env.row_height)) - \
                layout.get_pixel_size()[1]

            cr.translate(xOffset, yOffset)


def drawDay(cr, env, date):
    with drawing.restoring_transform(cr):
        yOffset = env.safety + \
            ((date.day - 1) * env.row_height) + (1.0 * env.row_height)

        # translate to drawing point top left corner
        cr.translate(0.0, yOffset)

        # fill box if weekend
        if(date.isoweekday() >= 6):
            cr.set_source_rgba(0.90, 0.90, 0.90, 1.0)
            cr.rectangle(0, 0, env.row_width, env.row_height)
            cr.fill()

        # stroke the box
        cr.set_source_rgba(0, 0, 0, 1.0)
        cr.set_line_width(env.line_width)
        cr.rectangle(0, 0, env.row_width, env.row_height)
        cr.stroke()

        daySize = math.floor(env.row_height * 0.25)
        numberSize = math.floor(env.row_height * 0.5)
        yOffset = (env.row_height* 0.25) / 2

        dayFont = Pango.FontDescription("%s %s" % (env.font, daySize)) # day text is way smaller than number
        numberFont = Pango.FontDescription("%s %s" % (env.font, numberSize))

        style = "%a" # by default abbreviated because need of space!
        # if(env.abbreviate or env.abbreviate_all):
        #     style = "%a"
        weekdayString = "%s" % (date.strftime(style))

        with drawing.restoring_transform(cr):
            with drawing.layout(cr) as numberLayout:
                numberLayout.set_font_description(numberFont)
                numberLayout.set_text("%s" % date.day, -1) # Number

                xOffset = math.ceil((env.row_width / 8)  - (numberLayout.get_pixel_size()[0] / 2))
                # yOffset = math.floor((env.row_height - numberSize - daySize - (daySize/2)) / 2)
                numberOffset = yOffset + ((numberSize - numberLayout.get_pixel_size()[1])/2)
                cr.translate(xOffset, numberOffset)

        with drawing.restoring_transform(cr):
            with drawing.layout(cr) as dayLayout:
                dayLayout.set_font_description(dayFont)
                dayLayout.set_text(weekdayString, -1) # Weekday

                xOffset = math.ceil((env.row_width / 8)  - (dayLayout.get_pixel_size()[0] / 2))
                yOffset = env.row_height - yOffset - daySize
                cr.translate(xOffset, yOffset)

File: src/layout/oneyear.rs
use chrono::{Datelike, NaiveDate};

use crate::calendar;
use crate::canvas::{Canvas, Font};
use crate::config::{CM, Config, Orientation, PageSetup};
use crate::style::Color;

pub struct OneYearLayout;

impl super::Layout for OneYearLayout {
    fn orientation(&self) -> Orientation {
        Orientation::Landscape
    }

    fn draw(&self, canvas: &Canvas, config: &Config, page: &PageSetup) {
        let font_size = 6.0;
        let line_width = 0.01 * CM;
        // 12 columns with 13 margins (safety between each column + edges)
        let row_width = (page.width - 13.0 * page.margin) / 12.0;
        let row_height = (page.height - 2.0 * page.margin) / 32.0;

        let mut date = NaiveDate::from_ymd_opt(config.year, config.month, 1).unwrap();

        for col in 0..12 {
            canvas.with_save(|c| {
                c.translate(
                    page.margin + (col as f64 * row_width) + (col as f64 * page.margin),
                    0.0,
                );

                // Month title
                draw_month_title(c, config, page, date, row_width, row_height);

                // Days
                let starting_month = date.month();
                while date.month() == starting_month {
                    draw_day(
                        c, config, date, row_width, row_height, font_size, line_width, page,
                    );
                    date = date.succ_opt().unwrap();
                }
            });
        }
    }
}

fn draw_month_title(
    canvas: &Canvas,
    config: &Config,
    page: &PageSetup,
    date: NaiveDate,
    row_width: f64,
    row_height: f64,
) {
    canvas.with_save(|c| {
        let size = (row_height * 0.9).ceil();
        let font = Font::new(config.heading_font(), size);

        // Always abbreviated and uppercased
        let month_str = date.format("%b").to_string().to_uppercase();

        let m = c.measure_text(&month_str, &font);
        let x_offset = (row_width - m.width) / 2.0;
        let y_offset = (page.margin + row_height) - m.height;

        c.draw_text(
            &month_str,
            x_offset,
            y_offset,
            &font,
            Color::rgb(0.0, 0.0, 0.0),
        );
    });
}

fn draw_day(
    canvas: &Canvas,
    config: &Config,
    date: NaiveDate,
    row_width: f64,
    row_height: f64,
    _font_size: f64,
    line_width: f64,
    page: &PageSetup,
) {
    canvas.with_save(|c| {
        let y_offset = page.margin + (date.day0() as f64 * row_height) + row_height;
        c.translate(0.0, y_offset);

        // Weekend background
        if calendar::is_weekend(date) {
            c.fill_rect(
                0.0,
                0.0,
                row_width,
                row_height,
                Color::rgb(0.90, 0.90, 0.90),
            );
        }

        // Bottom line (not full rectangle stroke, matching Python)
        c.draw_line(
            0.0,
            row_height,
            row_width,
            row_height,
            Color::rgb(0.0, 0.0, 0.0),
            line_width,
        );

        // Text sizes based on row height
        let number_size = (row_height * 0.4).floor();
        let day_size = (row_height * 0.25).floor();

        let number_font = Font::new(&config.font, number_size);
        let day_font = Font::new(&config.font, day_size);

        let weekday_str = date.format("%a").to_string();

        // Day number
        let day_str = date.day().to_string();
        let number_m = c.measure_text(&day_str, &number_font);
        let number_y = (row_height - number_size) / 2.0;

        canvas.with_save(|c| {
            c.translate(0.0, number_y);
            c.draw_text(&day_str, 0.0, 0.0, &number_font, Color::rgb(0.0, 0.0, 0.0));
        });

        // Weekday letter (first character)
        let weekday_char = &weekday_str[..1];
        let x_offset = number_m.width * 1.025;
        let day_y = (row_height - number_size) / 2.0 + (day_size * 0.8);

        canvas.with_save(|c| {
            c.translate(x_offset, day_y);
            c.draw_text(weekday_char, 0.0, 0.0, &day_font, Color::rgb(0.0, 0.0, 0.0));
        });
    });
}
